src/Common/UnreleasableLRUCache.h

```diff
@@ -0,0 +1,283 @@
+        queue.splice(queue.end(), queue, cell.queue_iterator);
+
+        return cell.value;
+    }
+
+    bool setImpl(const Key & key, const MappedPtr & mapped, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock)
+    {
+        auto [it, inserted] = cells.emplace(std::piecewise_construct,
+            std::forward_as_tuple(key),
+            std::forward_as_tuple());
+
+        Cell & cell = it->second;
+
+        if (inserted)
+        {
+            auto weight = mapped ? weight_function(*mapped) : 0;
+            if (!removeOverflow(weight))
+            {
+                // cannot insert this new value
+                cells.erase(it);
+                return false;
+            }
+
+            try
+            {
+                cell.queue_iterator = queue.insert(queue.end(), key);
+            }
+            catch (...)
+            {
+                cells.erase(it);
+                throw;
+            }
+        }
+        else
+        {
+            if (evict_policy.canRelease(*cell.value) != CacheEvictStatus::CAN_EVITCT)
+            {
+                // the old value is not releasable
+                return false;
+            }
+            evict_policy.release(*cell.value);
+            current_size -= cell.size;
+            queue.splice(queue.end(), queue, cell.queue_iterator);
+        }
+
+        cell.value = mapped;
+        cell.size = cell.value ? weight_function(*cell.value) : 0;
+        current_size += cell.size;
+
+        removeOverflow(0);
+        return true;
+    }
+
+    // Or make your own implementation
+    virtual bool removeOverflow(size_t more_size)
+    {
+        size_t current_weight_lost = 0;
+        size_t queue_size = cells.size();
+
+        auto key_it = queue.begin();
+        while ((current_size + more_size > max_size || (max_elements_size != 0 && queue_size > max_elements_size)) && (queue_size > 1)
+                && key_it != queue.end())
+        {
+            const Key & key = *key_it;
+
+            auto it = cells.find(key);
+            if (it == cells.end())
+            {
+                LOG_ERROR(&Poco::Logger::get("UnreleasableLRUCache"), "UnreleasableLRUCache became inconsistent. There must be a bug in it.");
+                abort();
+            }
+
+            const auto & cell = it->second;
+            auto cache_evict_status = evict_policy.canRelease(*(cell.value));
+            if (cache_evict_status == CacheEvictStatus::CAN_EVITCT)
+            {
+                evict_policy.release(*(cell.value));
+                current_size -= cell.size;
+                current_weight_lost += cell.size;
+
+                cells.erase(it);
+                key_it = queue.erase(key_it);
+                --queue_size;
+            }
+            else if (cache_evict_status == CacheEvictStatus::SKIP_EVICT)
+            {
+                key_it++;
+                continue;
+            }
+            else if (cache_evict_status == CacheEvictStatus::TERMINATE_EVICT)
```

Comment:
I cannot find where in code this `TERMINATE_EVICT` is set. Looks unused.

Reply:
It is for potential case that we want to stop the evict iteration when we meet the first unreleasable element

Reply:
Ok, but if it is needed better to consider it in code now, because if there is not, I would rather get rid of this enum and make one bool flag instead as now only 2 states a needed.